### PR TITLE
[RDY] Refactor InputResult out of AbstractSpoofaxResult hierarchy

### PR DESCRIPTION
**Pull request by Balletie**
The benefits of this PR:
1. `InputResult` has no accessor for an AST. The AST accessor method has been removed from `ISpoofaxResult` and moved to a subinterface `ISpoofaxTermResult`.
2. The `kleisliCompose` method now respects Liskov substitution principle: the output of the first function may be a subtype of the input argument of the second function.
3. Because `InputResult` has no AST accessor anymore, the ParseResult, AnalyzeResult and TransformResult are in separate class hierarchies, and have as common superclass `ISpoofaxTermResult`. And because the `kleisliCompose` method supports substitution, any subclass of `ISpoofaxTermResult` can be given to the `EvaluateFunction` as argument.
4. This in turn means that there is no more need for a `PEvalFunction` or `AEvalFunction`: there is now only one `EvaluateFunction` that accepts an `ISpoofaxTermResult` as its input.
5. `IEvaluationStrategy` has been simplified because of this, by accepting an `IStrategoTerm` instead of a `ParseResult` or `AnalyzeResult`.

### CODE REVIEW

File: org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/functions/EvaluateFunctionTest.java [hunk 1]
```diff
@@ -276,4 +302,24 @@ public void testEvaluateException() throws MetaborgException {
...
     }
 
+    /**
+     * Test throwing an exception when there is no {@link ShellFacet}.
+     *
+     * @throws MetaborgException
+     *             on unexpected Spoofax exceptions
+     */
+    @Test
+    public void testAbsenShellFacetException() throws MetaborgException {
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/test/java/org/metaborg/spoofax/shell/functions/EvaluateFunctionTest.java]: Absent :P

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/EvaluateFunction.java [hunk 2]
```diff
@@ -49,14 +50,18 @@ public AEvalFunction(Map<String, IEvaluationStrategy> evaluationStrategies,
...
+            return FailOrSuccessResult.failed(a);
+        }
         ShellFacet facet = context.language().facet(ShellFacet.class);
         if (facet == null) {
             throw new MetaborgException("Cannot find the shell facet.");
         }
 
         IEvaluationStrategy evalStrategy = evaluationStrategies.get(facet.getEvaluationMethod());
-        IStrategoTerm result = evalStrategy.evaluate(a, context);
+        IStrategoTerm result = evalStrategy.evaluate(a.ast().get(), context);
```
Comment (Unia) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/functions/EvaluateFunction.java]: Are you sure there is always an ast to `get()`?

File: org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/AbstractSpoofaxResult.java [hunk 3]
```diff
@@ -2,26 +2,20 @@
...
 /**
  * Represents an {@link AbstractSpoofaxResult} as returned by the {@link IReplCommand}.
  * Wraps Spoofax {@link IUnit} of various types.
  * @param <T> the wrapped subtype of {@link IUnit}
  */
-public abstract class AbstractSpoofaxResult<T extends IUnit> implements ISpoofaxResult<IUnit> {
-    private final IStrategoCommon common;
-    private final T unit;
+public abstract class AbstractSpoofaxResult<T extends IUnit> implements ISpoofaxResult<T> {
+    private T unit;
```
Comment (Hjdskes) [inline org.metaborg.spoofax.shell.core/src/main/java/org/metaborg/spoofax/shell/output/AbstractSpoofaxResult.java]: Can be `final`.

**Comment by Hjdskes** [comment]
Nice work! :smile:

**Pull request merged**